prepared_cache.go

```diff
@@ -1,64 +1,121 @@
 package gocql
 
 import (
-	"github.com/gocql/gocql/internal/lru"
+	"encoding/hex"
 	"sync"
+
+	"github.com/gocql/gocql/internal/lru"
+	"github.com/minio/highwayhash"
 )
 
-const defaultMaxPreparedStmts = 1000
+const (
+	defaultMaxPreparedStmts = 1000
+	shardCount              = 256
+)
 
 // preparedLRU is the prepared statement cache
 type preparedLRU struct {
-	mu  sync.Mutex
-	lru *lru.Cache
+	shards [shardCount]*shard
+	key    []byte
+}
+
+type shard struct {
+	items *lru.Cache
+	lock  *sync.RWMutex
+}
+
+// newPreparedLRU creates and initializes a * preparedLRU instance.
+func newPreparedLRU(size int) *preparedLRU {
+	key, err := hex.DecodeString("000102030405060708090A0B0C0D0E0FF0E0D0C0B0A090807060504030201000")
+	if err != nil {
+		panic(err)
+	}
+	c := &preparedLRU{
+		key: key,
+	}
+	for i := 0; i < shardCount; i++ {
+		c.shards[i] = &shard{
+			items: lru.New(size),
+			lock:  new(sync.RWMutex),
+		}
+	}
+	return c
 }
 
 // Max adjusts the maximum size of the cache and cleans up the oldest records if
 // the new max is lower than the previous value. Not concurrency safe.
 func (p *preparedLRU) max(max int) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-
-	for p.lru.Len() > max {
-		p.lru.RemoveOldest()
+	for _, shard := range p.shards {
+		shard.lock.Lock()
+		for shard.items.Len() > max {
+			shard.items.RemoveOldest()
+		}
+		shard.items.MaxEntries = max
+		shard.lock.Unlock()
 	}
-	p.lru.MaxEntries = max
 }
 
 func (p *preparedLRU) clear() {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-
-	for p.lru.Len() > 0 {
-		p.lru.RemoveOldest()
+	for _, shard := range p.shards {
+		shard.lock.Lock()
+		if shard.items.Len() > 0 {
+			shard.items.RemoveOldest()
+		}
+		shard.lock.Unlock()
 	}
 }
 
 func (p *preparedLRU) add(key string, val *inflightPrepare) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	p.lru.Add(key, val)
+	shard := p.shard(key)
+	shard.lock.Lock()
+	shard.items.Add(key, val)
+	shard.lock.Unlock()
 }
 
 func (p *preparedLRU) remove(key string) bool {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	return p.lru.Remove(key)
+	shard := p.shard(key)
+	shard.lock.Lock()
+	ok := shard.items.Remove(key)
+	shard.lock.Unlock()
+	return ok
 }
 
 func (p *preparedLRU) execIfMissing(key string, fn func(lru *lru.Cache) *inflightPrepare) (*inflightPrepare, bool) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+	shard := p.shard(key)
+	shard.lock.Lock()
+	defer shard.lock.Unlock()
 
-	val, ok := p.lru.Get(key)
+	val, ok := shard.items.Get(key)
 	if ok {
 		return val.(*inflightPrepare), true
 	}
 
-	return fn(p.lru), false
+	return fn(shard.items), false
 }
 
 func (p *preparedLRU) keyFor(addr, keyspace, statement string) string {
 	// TODO: maybe use []byte for keys?
 	return addr + keyspace + statement
 }
+
+func (c *preparedLRU) get(key string) *inflightPrepare {
```

Comment:
Can we also keep it consistent as either `c *preparedLRU` or `p *preparedLRU`

Reply:
Sure we can but I think "alt 2" is currently the most promising solution. Simple and should also work now ;)

Reply:
Sorry I meant "alt 3", the one with a simple read/write lock.

Reply:
:+1: